enclave-tls/src/tls_wrappers/openssl/un_negotiate.c

```diff
@@ -11,8 +11,96 @@
 #include "openssl.h"
+#include "sgx_report.h"
+#include "sgx_quote_3.h"
+
+#ifdef SGX
+tls_wrapper_err_t enclave_verify_mrenclave(tls_wrapper_ctx_t *tls_ctx,
+					   attestation_evidence_t *evidence)
+{
+	uint8_t mrenclave[32];
+	sgx_quote3_t quote3;
+
+	memset(mrenclave, 0, sizeof(mrenclave));
+	memset(&quote3, 0, sizeof(quote3));
+
+	memcpy(&quote3, evidence->ecdsa.quote, sizeof(quote3));
+
+	if (!strcmp(evidence->type, "sgx_ecdsa")) {
+		memcpy(mrenclave, quote3.report_body.mr_enclave.m, sizeof(mrenclave));
+		if (memcmp(tls_ctx->etls_handle->config.enclave_info.mrenclave, mrenclave,
+			   sizeof(mrenclave)))
+			return -TLS_WRAPPER_ERR_INVALID;
+	} else if (!strcmp(evidence->type, "sgx_ecdsa_qve")) {
+		memcpy(mrenclave, quote3.report_body.mr_enclave.m, sizeof(mrenclave));
+		if (memcmp(tls_ctx->etls_handle->config.enclave_info.mrenclave, mrenclave,
+			   sizeof(mrenclave)))
+			return -TLS_WRAPPER_ERR_INVALID;
+	} else if (!strcmp(evidence->type, "sgx_la")) {
+		sgx_report_t *lareport = (sgx_report_t *)(evidence->la.report);
+		memcpy(mrenclave, lareport->body.mr_enclave.m, sizeof(mrenclave));
+		if (memcmp(tls_ctx->etls_handle->config.enclave_info.mrenclave, mrenclave,
+			   sizeof(mrenclave)))
+			return -TLS_WRAPPER_ERR_INVALID;
+	}
+
+	return TLS_WRAPPER_ERR_NONE;
+}
+
+tls_wrapper_err_t enclave_verify_mrsigner(tls_wrapper_ctx_t *tls_ctx,
+					  attestation_evidence_t *evidence)
+{
+	uint8_t mrsigner[32];
+	sgx_quote3_t quote3;
+
+	memset(mrsigner, 0, sizeof(mrsigner));
+	memset(&quote3, 0, sizeof(quote3));
+
+	memcpy(&quote3, evidence->ecdsa.quote, sizeof(quote3));
+
+	if (!strcmp(evidence->type, "sgx_ecdsa")) {
+		memcpy(mrsigner, quote3.report_body.mr_signer.m, sizeof(mrsigner));
+		if (memcmp(tls_ctx->etls_handle->config.enclave_info.mrsigner, mrsigner,
+			   sizeof(mrsigner)))
+			return -TLS_WRAPPER_ERR_INVALID;
+	} else if (!strcmp(evidence->type, "sgx_ecdsa_qve")) {
+		memcpy(mrsigner, quote3.report_body.mr_signer.m, sizeof(mrsigner));
+		if (memcmp(tls_ctx->etls_handle->config.enclave_info.mrsigner, mrsigner,
+			   sizeof(mrsigner)))
+			return -TLS_WRAPPER_ERR_INVALID;
+	} else if (!strcmp(evidence->type, "sgx_la")) {
+		sgx_report_t *lareport = (sgx_report_t *)evidence->la.report;
+		memcpy(mrsigner, lareport->body.mr_signer.m, sizeof(mrsigner));
+		if (memcmp(tls_ctx->etls_handle->config.enclave_info.mrsigner, mrsigner,
+			   sizeof(mrsigner)))
+			return -TLS_WRAPPER_ERR_INVALID;
+	}
+
+	return TLS_WRAPPER_ERR_NONE;
+}
+
+tls_wrapper_err_t enclave_user_callback(tls_wrapper_ctx_t *tls_ctx,
+					attestation_evidence_t *evidence)
+{
+	tls_wrapper_err_t ret;
+	ret = enclave_verify_mrenclave(tls_ctx, evidence);
+	if (ret != TLS_WRAPPER_ERR_NONE) {
+		ETLS_ERR("mrenclave does not match\n");
+		return ret;
+	}
+
+	ret = enclave_verify_mrsigner(tls_ctx, evidence);
+	if (ret != TLS_WRAPPER_ERR_NONE) {
+		ETLS_ERR("mrsigner does not match\n");
+		return ret;
+	}
+
+	return TLS_WRAPPER_ERR_NONE;
+}
+#endif
```

Comment:
The user callback for sgx evidence is a generic logic, so it should not be implemented in an instance.

Reply:
user callback is releated with verify callback. So you mean verify callback should not be implement in an instance as well?

Reply:
sgx-epid (if any) and sgx-ecdsa will share this logic. So it is necessary to abstract a common sgx logic to implement this callback, .e,g, write a sgx_evidence.c and implement this common callback sgx_verification_common_check(). Because sgx-* verifier instances always set SGX_CAP in config flags so the core layer can register sgx_verification_common_check().